### PR TITLE
rpc_dev

### PR DESCRIPTION
**Pull request by gtygo**
add two command function : maketx & signrawtx

### CODE REVIEW

File: commands/box/transaction/root.go [hunk 1]
```diff
@@ -95,6 +104,79 @@ to quickly create a Cobra  application.`,
...
+		To:      toAddr,
+		Amounts: amounts,
+		Fee:     uint64(fee),
+	}
+	conn, err := rpcutil.GetGRPCConn(getRPCAddr())
+	if err != nil {
+		fmt.Println(err)
+	}
+	defer conn.Close()
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: 10s may be better

File: commands/box/transaction/root.go [hunk 2]
```diff
@@ -59,12 +65,15 @@ func init() {
 				fmt.Println("sendtoaddress called")
 			},
 		},
+		&cobra.Command{
+			Use:   "maketx [from] [to1,to2...] [amounts1,amounts2...] [fee]",
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: should use gas price

File: commands/box/transaction/root.go [hunk 3]
```diff
@@ -59,12 +65,15 @@ func init() {
...
+			Short: "make unsigned transaction",
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: create unsigned transaction

File: commands/box/transaction/root.go (hunk 2 above)
Comment (Jerick26) [inline commands/box/transaction/root.go]: createtx

File: commands/box/transaction/root.go [hunk 4]
```diff
@@ -95,6 +104,79 @@ to quickly create a Cobra  application.`,
...
+	if err := types.ValidateAddr(append(toAddr, from)...); err != nil {
+		fmt.Println("validate address error", err)
+		return
+	}
+	amountStr := strings.Split(args[2], ",")
+	amounts := make([]uint64, 0)
+	for _, x := range amountStr {
+		tmp, err := strconv.Atoi(x)
+		if err != nil {
+			fmt.Println("Type conversion failed: ", err)
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: "invalid argument" may be better

File: commands/box/transaction/root.go [hunk 5]
```diff
@@ -95,6 +104,79 @@ to quickly create a Cobra  application.`,
...
+			return
+		}
+		amounts = append(amounts, uint64(tmp))
+	}
+
+	fee, err := strconv.Atoi(args[3])
+	if err != nil {
+		fmt.Println("Type conversion failed: ", err)
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: as above

File: commands/box/transaction/root.go [hunk 6]
```diff
@@ -95,6 +104,79 @@ to quickly create a Cobra  application.`,
...
+	conn, err := rpcutil.GetGRPCConn(getRPCAddr())
+	if err != nil {
+		fmt.Println(err)
+	}
+	defer conn.Close()
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	client := rpcpb.NewTransactionCommandClient(conn)
+	resp, err := client.MakeUnsignedTx(ctx, req)
+	tx := resp.Tx
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: ??

File: commands/box/transaction/root.go [hunk 7]
```diff
@@ -95,6 +104,79 @@ to quickly create a Cobra  application.`,
...
+
+	if err != nil {
+		fmt.Println("make transaction failed: ", err)
+		return
+	}
+	rawMsgs := resp.RawMsgs
+	if err != nil {
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: ??

File: commands/box/transaction/root.go [hunk 8]
```diff
@@ -95,6 +104,79 @@ to quickly create a Cobra  application.`,
...
+		fmt.Println("get raw message failed:", err)
+		return
+	}
+	fmt.Println("transaction: ", util.PrettyPrint(tx))
+	rawTxByte, err := tx.Marshal()
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: rawTxBytes

File: commands/box/transaction/root.go [hunk 9]
```diff
@@ -95,6 +104,79 @@ to quickly create a Cobra  application.`,
...
+	if err != nil {
+		fmt.Println("transaction marshal failed:", err)
+	}
+	rawTx := hex.EncodeToString(rawTxByte)
+	fmt.Println("raw transaction :", rawTx)
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: Output format must be identical.
fmt.Println("raw transaction: ", rawTx)

File: commands/box/transaction/root.go [hunk 10]
```diff
@@ -242,6 +324,80 @@ func decoderawtx(cmd *cobra.Command, args []string) {
 	fmt.Println(util.PrettyPrint(tx))
 }
 
+func signrawtx(cmd *cobra.Command, args []string) {
+	fmt.Println("signrawtx called")
+	if len(args) != 2 {
+		fmt.Println("Invalide argument number")
+		return
+	}
+	txByte, err := hex.DecodeString(args[0])
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: txBytes

File: commands/box/transaction/root.go [hunk 11]
```diff
@@ -59,12 +65,15 @@ func init() {
 				fmt.Println("sendtoaddress called")
 			},
 		},
+		&cobra.Command{
+			Use:   "maketx [from] [to1,to2...] [amounts1,amounts2...] [fee]",
+			Short: "make unsigned transaction",
+			Run:   maketx,
+		},
 		&cobra.Command{
 			Use:   "signrawtx [rawtx]",
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: parameters are not identical with function signrawtx

File: commands/box/transaction/root.go [hunk 12]
```diff
@@ -242,6 +324,80 @@ func decoderawtx(cmd *cobra.Command, args []string) {
...
+	tx := new(types.Transaction)
+	if err := tx.Unmarshal(txByte); err != nil {
+		fmt.Println("Unmarshal error: ", err)
+		return
+	}
+	wltMgr, err := wallet.NewWalletManager(walletDir)
+	if err != nil {
+		fmt.Println("get wallet manager error", err)
+	}
+	accI, ok := wltMgr.GetAccount(args[1])
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: args[1] is private key?
it is not secure!

File: commands/box/transaction/root.go [hunk 13]
```diff
@@ -242,6 +324,80 @@ func decoderawtx(cmd *cobra.Command, args []string) {
...
+	passphrase, err := wallet.ReadPassphraseStdin()
+	if err := accI.UnlockWithPassphrase(passphrase); err != nil {
+		fmt.Println("Fail to unlock account", err)
+		return
+	}
+	cmdReader := bufio.NewReader(os.Stdin)
+	vinAmount := len(tx.Vin)
+	rawMsgs := make([][]byte, 0)
+	for i := 0; i < vinAmount; i++ {
+		fmt.Println("Please enter rawmsg for this transaction: ")
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: raw msgs should be used as arguments of signrawtx

File: commands/box/transaction/root.go [hunk 14]
```diff
@@ -242,6 +324,80 @@ func decoderawtx(cmd *cobra.Command, args []string) {
...
+		cmdStr = strings.Trim(cmdStr, "\r\n")
+		rawMsgByte, err := hex.DecodeString(cmdStr)
+		if err != nil {
+			fmt.Println(err)
+			return
+		}
+		rawMsgs = append(rawMsgs, rawMsgByte)
+	}
+	sigHashes := make([]*crypto.HashType, 0, len(rawMsgs))
+	for _, msg := range rawMsgs {
```
Comment (Jerick26) [inline commands/box/transaction/root.go]: why need three for loops?

File: commands/box/transaction/root.go (hunk 1 above)
Comment (gtygo) [reply commands/box/transaction/root.go]: done

File: commands/box/transaction/root.go (hunk 3 above)
Comment (gtygo) [reply commands/box/transaction/root.go]: done

File: commands/box/transaction/root.go (hunk 8 above)
Comment (gtygo) [reply commands/box/transaction/root.go]: done

File: commands/box/transaction/root.go (hunk 7 above)
Comment (gtygo) [reply commands/box/transaction/root.go]: done

File: commands/box/transaction/root.go (hunk 12 above)
Comment (gtygo) [reply commands/box/transaction/root.go]: args[1] is address,not private key.

**Pull request closed without merging**